arches_lingo/views/api/concepts.py

```diff
@@ -0,0 +1,88 @@
+from http import HTTPStatus
+
+from django.core.paginator import Paginator
+from django.utils.decorators import method_decorator
+from django.utils.translation import gettext as _
+from django.views.generic import View
+
+from arches.app.models.system_settings import settings
+from arches.app.utils.decorators import group_required
+from arches.app.utils.response import JSONErrorResponse, JSONResponse
+
+from arches_lingo.models import VwLabelValue
+from arches_lingo.utils.concept_builder import ConceptBuilder
+
+
+@method_decorator(
+    group_required("RDM Administrator", raise_exception=True), name="dispatch"
+)
+class ConceptTreeView(View):
+    def get(self, request):
+        builder = ConceptBuilder()
+        data = {
+            "schemes": [builder.serialize_scheme(scheme) for scheme in builder.schemes]
+        }
+        return JSONResponse(data)
+
+
+@method_decorator(
+    group_required("RDM Administrator", raise_exception=True), name="dispatch"
+)
+class ValueSearchView(ConceptTreeView):
+    def get(self, request):
+        term = request.GET.get("term")
+        max_edit_distance = request.GET.get(
+            "maxEditDistance", self.default_sensitivity()
+        )
+        exact = request.GET.get("exact", False)
+        page_number = request.GET.get("page", 1)
+        items_per_page = request.GET.get("items", 25)
+
+        if exact:
+            concept_query = VwLabelValue.objects.filter(value=term).order_by(
+                "concept_id"
+            )
+        elif term:
+            try:
+                concept_query = VwLabelValue.objects.fuzzy_search(
+                    term, max_edit_distance
+                )
+            except ValueError as ve:
+                return JSONErrorResponse(
+                    title=_("Unable to perform search."),
+                    message=ve.args[0],
+                    status=HTTPStatus.BAD_REQUEST,
+                )
+        else:
+            concept_query = VwLabelValue.objects.all().order_by("concept_id")
+        concept_ids = concept_query.values_list("concept_id", flat=True).distinct()
+
+        paginator = Paginator(concept_ids, items_per_page)
```

Comment:
That way, this code gets reduced to something like:

```
        if not concept_ids:
            # No results: don't bother building the concept tree.
            return JSONResponse([])

        paginator = Paginator(concept_ids, items_per_page)
        page = paginator.get_page(page_number)

        builder = ConceptBuilder()

        return JSONResponse([
            builder.serialize_concept(str(concept_uuid), parents=True, children=False)
            for concept_uuid in page
        ])
```

Reply:
Yeah, as I mentioned above, it would be a pessimization to evaluate the queryset before the paginator slices it. We need delegate that to the paginator.